aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -679,6 +682,167 @@ impl AptosVM {
+                        .get(value.name.as_ident_str().as_str())
+                        .ok_or_else(|| {
+                            Self::metadata_validation_error("Invalid resource_group attribute")
+                        })?
+                } else {
+                    return Err(Self::metadata_validation_error("No such resource_group"));
+                };
+
+                if !scope.are_equal_module_ids(&module_id, &value_module_id) {
+                    return Err(
+                        PartialVMError::new(StatusCode::LINKER_ERROR).finish(Location::Undefined)
+                    );
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Validate resource group metadata on a single module
+    /// * Extract the resource group metadata
+    /// * Verify all changes are compatible upgrades
+    /// * Return any new members to validate correctness and all groups to assist in validation
+    fn validate_module_and_extract_new_entries<S: MoveResolverExt>(
+        session: &mut SessionExt<S>,
+        module: &CompiledModule,
+    ) -> VMResult<(
+        BTreeMap<String, ResourceGroupScope>,
+        BTreeMap<String, StructTag>,
+    )> {
+        let (new_groups, mut new_members) =
+            if let Some(metadata) = aptos_framework::get_module_metadata(module) {
+                Self::extract_resource_group_metadata(&metadata)?
+            } else {
+                (BTreeMap::new(), BTreeMap::new())
+            };
+
+        let (original_groups, original_members) =
+            Self::extract_resource_group_metadata_from_module(session, &module.self_id())?;
+
+        for (member, value) in original_members {
+            // We don't need to re-validate new_members above.
+            if Some(&value) != new_members.remove(&member).as_ref() {
+                return Err(Self::metadata_validation_error(
+                    "Invalid change in resource_group_member",
+                ));
+            }
+        }
+
+        for (group, value) in original_groups {
+            // We need groups in case there's cross module dependencies
+            if let Some(new_value) = new_groups.get(&group) {
+                if value.is_less_strict(new_value) {
+                    return Err(Self::metadata_validation_error(
+                        "Invalid change in resource_group",
+                    ));
+                }
+            } else {
+                return Err(Self::metadata_validation_error(
+                    "Invalid change in resource_group",
+                ));
+            }
+        }
+
+        Ok((new_groups, new_members))
+    }
+
+    /// Given a module id extract all resource group metadata
+    fn extract_resource_group_metadata_from_module<S: MoveResolverExt>(
+        session: &mut SessionExt<S>,
+        module_id: &ModuleId,
+    ) -> VMResult<(
+        BTreeMap<String, ResourceGroupScope>,
+        BTreeMap<String, StructTag>,
+    )> {
+        let metadata = session
```

Comment:
Why is this not going over the VM? The VM's loader should cache the full metadata (the new design will consider this), and at least already now caches the deserialized module. 

This does not need to block this PR but something which should be revisted. Perhaps add a TODO? In general the needs for metadata should go into the design of the new loader cache.